### PR TITLE
ConnectionMetadata is now ConnectionItems

### PR DESCRIPTION
**Pull request by davidfowl**
Reacting to https://github.com/aspnet/KestrelHttpServer/pull/2395

### CODE REVIEW

File: samples/SocialWeather/PersistentConnectionLifeTimeManager.cs [hunk 1]
```diff
@@ -23,9 +23,9 @@ public PersistentConnectionLifeTimeManager(FormatterResolver formatterResolver)
 
         public void OnConnectedAsync(ConnectionContext connection)
         {
-            connection.Features.Get<IConnectionMetadataFeature>().Metadata["groups"] = new HashSet<string>();
+            connection.Items["groups"] = new HashSet<string>();
```
Comment (JamesNK) [inline samples/SocialWeather/PersistentConnectionLifeTimeManager.cs]: Can Items be shared with other pieces of code? Should there be a prefix on it so make it less likely to conflict with someone else who might add data with a `"groups"` key? e.g. `"socialweather-groups"`.

Also these keys should be `const` fields. It's just a sample but try to encourage good practice though leadership :)

File: samples/SocketsSample/EndPoints/MessagesEndPoint.cs [hunk 2]
```diff
@@ -19,7 +19,7 @@ public class MessagesEndPoint : EndPoint
         {
             Connections.Add(connection);
 
-            await Broadcast($"{connection.ConnectionId} connected ({connection.Features.Get<IConnectionMetadataFeature>().Metadata[ConnectionMetadataNames.Transport]})");
+            await Broadcast($"{connection.ConnectionId} connected ({connection.Items[ConnectionMetadataNames.Transport]})");
```
Comment (JamesNK) [inline samples/SocketsSample/EndPoints/MessagesEndPoint.cs]: So much nicer 👼

File: src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs [hunk 3]
```diff
@@ -66,9 +68,9 @@ public DefaultConnectionContext(string id, IDuplexPipe transport, IDuplexPipe ap
 
         public ClaimsPrincipal User { get; set; }
 
-        public IDictionary<object, object> Metadata { get; set; } = new ConnectionMetadata();
+        public override IDictionary<object, object> Items { get; set; } = new ConnectionMetadata();
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs]: Still need `ConnectionMetadata`?

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs]: That's correct. The base class does nothing.

File: samples/SocialWeather/PersistentConnectionLifeTimeManager.cs (hunk 1 above)
Comment (davidfowl) [reply samples/SocialWeather/PersistentConnectionLifeTimeManager.cs]: In general yea, but this is just a crappy sample 😄

File: src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs (hunk 3 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs]: I see `ConnectionMetadata` makes it safe to get a value that doesn't exist.

```
// Replace the indexer with one that returns null for missing values
object IDictionary<object, object>.this[object key]
{
    get
    {
        if (Items.TryGetValue(key, out var value))
        {
            return value;
        }
        return null;
    }
    set { Items[key] = value; }
}
```

If we're doing it here, should this be the default value of `Items` in Kestrel?

At the very least I don't think we need to override Items on DefaultConnectionContext.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs]: That sounds like a fine change to make but right now it's abstract in kestrel.

Comment (JamesNK) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs]: Ah ok, I didn't realize it was an abstract class. Carry on then.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs]: You're talking about making it non virtual right?

Comment (JamesNK) [reply src/Microsoft.AspNetCore.Sockets.Abstractions/DefaultConnectionContext.cs]: No, I just didn't realize `ConnectionContext` was abstract.

**Pull request merged**